### PR TITLE
Enforce specific HTML elements and single spaces in notes

### LINKED ISSUE #2799: Add documentation and linter for escaping HTML in the `notes` property
Currently we don't use the linter to automatically check for HTML elements in the data files, nor do we have documentation on escaping HTML elements in the data files. ([The docs do mention that you _can_ use `a` and `code`, but they say nothing about any other elements](https://github.com/mdn/browser-compat-data/blob/28f01acf62c205d53415c3f5da26a5847e90bdc0/schemas/compat-data-schema.md#notes))

Examples:

- https://github.com/mdn/browser-compat-data/pull/2776#discussion_r216154175
- https://github.com/mdn/browser-compat-data/pull/2770#discussion_r216065017
- https://github.com/mdn/browser-compat-data/pull/2211#discussion_r197425652

An unescaped `<button>` element actually made it into the `input type="submit"` data (it was mentioned in review but never fixed):

https://github.com/mdn/browser-compat-data/blob/28f01acf62c205d53415c3f5da26a5847e90bdc0/html/elements/input/submit.json#L24

`<a href='url-not-important'>persists the dynamic disabled state</a> of a <code><button></code> across page loads`

I didn't really think about this with my explorer app (my bad, admittedly), so theoretically this could be used to inject scripts into a site (or editor, e.g. VS Code) that uses the data: 

screen shot 2018-09-10 at 12 08 50 pm

MDN handles it better, but oddly adds a closing tag as well?

screen shot 2018-09-10 at 12 09 47 pm

Two main things we should do here:

- Add a linter rule to fail on cases like this (excluding `a` and `code`, maybe others I'm forgetting?), because using regex to parse HTML is _so_ easy :P 
- Add docs that mention the best way to escape HTML elements when using them in the BCD notes (e.g. replacing `<code><button></code>` with `<code>&lt;button&gt;</code>`?)

Alternatively, we could just make sure the docs warn third party implementors that we don't guarantee the safety of the data files and that they should only allow `a` and `code` in notes/descriptions, and no other HTML elements.

### LINKED ISSUE #3985: Lint for broken HTML
As far as I can tell, this: https://github.com/mdn/browser-compat-data/pull/3978 could have been caught by a linter checking for valid HTML (using e.g. https://validator.w3.org/).

### LINKED ISSUE #4033: Lint for unnnecessary attributes in markup (e.g. inside notes)
As per https://github.com/mdn/browser-compat-data/pull/3981#issuecomment-488245877

We ought to update the linter to flag unneccessary attributes (like title attributes in notes); they aren't useful and mean more of a buden if we try to localize BCD in the future.

### LINKED ISSUE #4829: Single/double spacing in notes?
With all the styling we're incorporating, I'm wondering if we should enforce single or double spacing in between sentences in notes, to maintain a more consistent writing style.  Since double-spaces get rendered as single spaces by default in browsers anyways, I'm thinking we should stick to single spacing, but I figured I'd open up a discussion here!

### PR DESCRIPTION
**Pull request by queengooborg**
This PR introduces a brand new linter that enforces the HTML elements within the notes.  A while ago, #4259 was opened that presents a regex-based solution, however it was closed as we wanted to use a parser-based solution instead.  That's what this PR does.  This uses the `html-validate` package to validate the HTML and ensure there's no broken HTML first.  Then, this uses the `node-html-parser` package to parse the HTML and check to ensure the elements within the notes are only one of the five allowed element types.  This also checks the attributes on tags, and ensures that `<a>` tags only have an `href` attribute, while all other tags have no attributes.  Finally, this enforces single-spacing in notes for consistency.

This PR fixes #2799, fixes #3985, fixes #4033, and fixes #4829.  ~~This linter will fail for the time being, as we currently have conflicting data.~~

### CODE REVIEW

**Review by alshamiri5 — APPROVED** [review]

**Comment by foolip** [comment]
@queengooborg does this lint pass with the recent note fix PR landed?

**Comment by queengooborg** [comment]
It does, yes, so this is now ready to be merged!

File: test/linter/test-notes.js [hunk 1]
```diff
@@ -0,0 +1,217 @@
+'use strict';
+const chalk = require('chalk');
+const parser = require('node-html-parser');
+const { HtmlValidate } = require('html-validate');
+const { VALID_ELEMENTS } = require('../utils.js');
```
Comment (foolip) [inline test/linter/test-notes.js:5]: `VALID_ELEMENTS` is only used in this lint rule, can we just put it here and not in utils? It seems unlikely we'll use it anywhere else.

File: test/linter/test-notes.js [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+const validator = new HtmlValidate();
+
+/**
+ * @typedef {import('../../types').Identifier} Identifier
+ *
+ * @typedef {'disallowed' | 'attrs' | 'attrs_a' | 'invalid' | 'doublespace'} HTMLErrorType
+ *
+ * @typedef {object} HTMLError
+ * @property {HTMLErrorType} errortype The type of error
+ * @property {string} relPath The identifier of the feature
```
Comment (foolip) [inline test/linter/test-notes.js:16]: What do the `relPath` variable names mean? Relative path? If so, relative to what?

File: test/linter/test-notes.js [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+ * @returns {void}
+ */
+const validateHTML = (string, browser, relPath, errors) => {
+  const report = validator.validateString(string, {
+    rules: {
+      'attr-quotes': 'off',
+    },
+  });
+  if (report.valid) {
+    // If HTML is valid, ensure we're only using valid elements
```
Comment (foolip) [inline test/linter/test-notes.js:83]: Can we unconditionally do this validation step? There's no input string that doesn't parse into *something*, valid or not.

File: test/linter/test-notes.js [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+ * @typedef {object} HTMLError
+ * @property {HTMLErrorType} errortype The type of error
+ * @property {string} relPath The identifier of the feature
+ * @property {string} browser The browser where the error was found
+ * @property {string} tag The specific HTML tag
+ * @property {string[]} messages Messages to describe the error
+ */
+
+/**
+ * Test a specific HTML tag (node) in notes for issues
```
Comment (foolip) [inline test/linter/test-notes.js:23]:
```suggestion
 * Recursively test a DOM node for valid elements
```

File: test/linter/test-notes.js [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+ *
+ * @param {object} node The HTML node to test
```
Comment (foolip) [inline test/linter/test-notes.js:25]:
```suggestion
 * @param {object} node The DOM node to test
```

File: test/linter/test-notes.js [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+const validateHTML = (string, browser, relPath, errors) => {
+  const report = validator.validateString(string, {
+    rules: {
+      'attr-quotes': 'off',
+    },
+  });
+  if (report.valid) {
+    // If HTML is valid, ensure we're only using valid elements
+    let data = parser.parse(string);
+    testNode(data, browser, relPath, errors);
```
Comment (foolip) [inline test/linter/test-notes.js:85]:
```suggestion
    const root = parser.parse(string);
    testNode(root, browser, relPath, errors);
```

File: test/linter/test-notes.js [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+  if (node.nodeType == 1) {
+    const tag = node.tagName?.toLowerCase();
+    if (tag && !VALID_ELEMENTS.includes(tag))
+      errors.push({
+        type: 'disallowed',
+        relPath,
+        browser,
+        tag,
+      });
+    if (tag !== 'a' && Object.entries(node.attributes).length !== 0) {
```
Comment (foolip) [inline test/linter/test-notes.js:41]: How about putting the bits for `a` in a `if (tag === 'a')` and making this an else-branch? That makes it entirely obvious that both code paths can't be taken.

And perhaps `const attributeCount = Object.entries(node.attributes).length;` first for reuse?

File: test/linter/test-notes.js [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        type: 'attrs',
+        relPath,
+        browser,
+        tag,
+      });
+    }
+    if (
+      tag === 'a' &&
+      (Object.entries(node.attributes).length !== 1 ||
+        Object.entries(node.attributes)[0][0] !== 'href')
```
Comment (foolip) [inline test/linter/test-notes.js:52]: This bit you can rewrite to `'href' in node.attributes`.

File: test/linter/test-notes.js [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+const testNotes = filename => {
+  /** @type {Identifier} */
+  const data = require(filename);
+
+  /** @type {HTMLError[]} */
+  const errors = [];
+
+  processData(data, errors);
+
+  if (errors.length) {
```
Comment (foolip) [inline test/linter/test-notes.js:171]: How about an early return to reduce the amount of indentation that follows?

```suggestion
  if (!errors.length) {
    return false;
  }
```

**Review by foolip — COMMENTED** [review]
Looking good! Can you also add tests for this that exercise each kind of error?

File: test/linter/test-notes.js [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+
+/**
+ * Process the data for notes errors
+ *
+ * @param {Identifier} data The data to test
+ * @param {HTMLError[]} errors The array to push errors to
+ * @param {string} [relPath] The identifier of the feature
+ * @returns {void}
+ */
+const processData = (data, errors, relPath) => {
```
[the remark this answers is not in the document]
Comment (queengooborg) [reply test/linter/test-notes.js]: Probably so, but that goes for all of our linters.  Could we hold off on refactoring this until the next-gen linter system, which should resolve a lot of the issues here?

**Comment by queengooborg** [comment]
@foolip This lint now passes so this is ready for merging!

**Review by foolip — APPROVED** [review]

**Comment by foolip** [comment]
@queengooborg did you see my question about tests for this, do you think that would make sense?

**Comment by queengooborg** [comment]
I definitely think tests would make a lot of sense, yes -- before digging into unittesting though, I'm hoping to get the next-gen linter implemented first, since i imagine that any existing unittests will need to be redone as well.

**Comment by queengooborg** [comment]
@foolip does the above "plan" make sense, and are you down to merge this linter before our data breaks again?

**Pull request merged**